src/Sylius/Component/Core/Inventory/Operator/OrderInventoryOperator.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Inventory\Operator;
+
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemInterface;
+use Sylius\Component\Core\OrderCheckoutStates;
+use Sylius\Component\Core\OrderPaymentStates;
+use Webmozart\Assert\Assert;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class OrderInventoryOperator implements OrderInventoryOperatorInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function updateInventoryState(OrderInterface $order) {
+        /** @var OrderItemInterface $orderItem */
+        foreach ($order->getItems() as $orderItem) {
+            $variant = $orderItem->getVariant();
+
+            if (!$variant->isTracked()) {
+                continue;
+            }
+
+            $variantOnHandQuantity = $variant->getOnHand();
+            $variantOnHoldQuantity = $variant->getOnHold();
+            $orderItemQuantity = $orderItem->getQuantity();
+
+            if (OrderPaymentStates::STATE_PAID === $order->getPaymentState()) {
+                if (OrderInterface::STATE_CANCELLED === $order->getState()) {
+                    $variant->setOnHand($variantOnHandQuantity + $orderItemQuantity);
+                    continue;
+                }
+
+                Assert::greaterThanEq(
+                    ($variantOnHoldQuantity - $orderItemQuantity),
+                    0,
+                    sprintf('Not enough units to decrease on hold quantity from the inventory of a variant "%s".', $variant->getName())
+                );
+
+                Assert::greaterThanEq(
+                    ($variantOnHandQuantity - $orderItemQuantity),
+                    0,
+                    sprintf('Not enough units to decrease on hand quantity from the inventory of a variant "%s".', $variant->getName())
+                );
+
+                $variant->setOnHand($variantOnHandQuantity - $orderItemQuantity);
+                $variant->setOnHold($variantOnHoldQuantity - $orderItemQuantity);
+                continue;
+            }
+
+            if (OrderInterface::STATE_CANCELLED === $order->getState()) {
+                Assert::greaterThanEq(
+                    ($variantOnHoldQuantity - $orderItemQuantity),
+                    0,
+                    sprintf('Not enough units to decrease the inventory of a variant "%s".', $variant->getName())
+                );
+
+                $variant->setOnHold($variantOnHoldQuantity - $orderItemQuantity);
+                continue;
+            }
+
+            if (OrderCheckoutStates::STATE_COMPLETED === $order->getCheckoutState()) {
+                $variant->setOnHold($variantOnHoldQuantity + $orderItemQuantity);
+                continue;
+            }
+        }
+    }
```

Comment:
This method should be splitted into 2: `hold` and `release`.

`hold` method will have the same logic as previous `decrease` in `OnHoldQuantityUpdater` (see [this](https://github.com/Sylius/Sylius/pull/6192/files#diff-5a43679c8e7edcb09593507831f6163aL42)) and should be called on checkout complete transition (no need to check the order state in the method).

`release` method should be called on before cancel transition and have the following logic: If order payment state is `paid` then `onHand++` otherwise `onHold--`.